tempesta_fw/http_frame.c

```diff
@@ -1655,27 +1655,27 @@ tfw_h2_frame_process(void *c, TfwFsmData *data)
 {
 	int r;
 	bool postponed;
-	unsigned int unused, curr_tail;
+	unsigned int unused;
 	TfwFsmData data_up = {};
 	TfwH2Ctx *h2 = tfw_h2_context(c);
 	struct sk_buff *nskb = NULL, *skb = data->skb;
-	unsigned int parsed = 0, off = data->off, tail = data->trail;
+	unsigned int parsed = 0, off = data->off, data_left = data->len;
 
 	BUG_ON(off >= skb->len);
-	BUG_ON(tail >= skb->len);
+	BUG_ON(off + data_left > skb->len);
 
 next_msg:
 	postponed = false;
 	ss_skb_queue_tail(&h2->skb_head, skb);
-	r = ss_skb_process(skb, off, tail, tfw_h2_frame_recv, h2, &unused,
+	r = ss_skb_process(skb, off, data_left, tfw_h2_frame_recv, h2, &unused,
 			   &parsed);
-
-	curr_tail = off + parsed + tail < skb->len ? 0 : tail;
-	if (r >= T_POSTPONE && ss_skb_chop_head_tail(NULL, skb, off, curr_tail))
+	if (r >= T_POSTPONE &&
+	    ss_skb_chop_head_tail(NULL, skb, off, skb->len - off - data_left))
 	{
 		r = T_DROP;
 		goto out;
 	}
+	data_left -= parsed;
```

Comment:
It seems that we also should set `data_up.len` to `parsed` value below - before `tfw_http_msg_process_generic()` call.

Reply:
Thanks for pointing this out. I'll dig deeper into the `data_up` processing.

Reply:
Added.

Reply:
There were changes that removed the need to handle heading and trailing parts of messages. It's no longer required to track amount data available in the instances of `TfwFsmData`, so the code mentioned in this discussion thread was removed.